src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -2128,6 +2128,91 @@ public static bool IsDefiniteCastTypeContext(this SyntaxTree syntaxTree, int pos
+                case SyntaxKind.NotPattern:
+                case SyntaxKind.ParenthesizedPattern:
+                case SyntaxKind.ParenthesizedExpression:
+                case SyntaxKind.PositionalPatternClause:
+                case SyntaxKind.PropertyPatternClause:
+                case SyntaxKind.RecursivePattern:
+                case SyntaxKind.RelationalPattern:
+                case SyntaxKind.SlicePattern:
+                case SyntaxKind.Subpattern:
+                case SyntaxKind.TypePattern:
+                case SyntaxKind.VarPattern:
+
+                // When we encounter the case of `X.$$.A`, we get the following
+                // - SimpleMemberAccessExpression for complex is patterns (within `IsPatternExpressionSyntax`)
+                case SyntaxKind.SimpleMemberAccessExpression:
+                // - QualifiedName for simple is patterns (within `BinaryExpressionSyntax(SyntaxKind.IsExpression)`)
+                case SyntaxKind.QualifiedName:
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        public static bool IsConstantPatternContext(this SyntaxTree syntaxTree, SyntaxToken tokenOnLeftOfPosition)
+        {
+            var node = tokenOnLeftOfPosition.Parent;
+            if (node is null)
+                return false;
+
+            var previousNode = node;
+            node = node?.Parent;
+            while (node is not null)
+            {
+                var kind = node.Kind();
+                switch (node.Kind())
+                {
+                    // nameof(...) expressions are parsed as InvocationExpression
+                    // And regular invocations are not available in a constant pattern context
+                    // so we are safe to always exit
+                    case SyntaxKind.InvocationExpression:
+                        return false;
+
+                    case SyntaxKind.Subpattern:
+                        {
+                            var subpattern = (SubpatternSyntax)node;
+                            if (subpattern.Pattern != previousNode)
+                                return false;
+
+                            break;
+                        }
+
+                    case SyntaxKind.IsPatternExpression:
+                        {
+                            var isPattern = (IsPatternExpressionSyntax)node;
+                            return isPattern.Pattern == previousNode;
+                        }
+                    case SyntaxKind.IsExpression:
+                        {
+                            var expression = (BinaryExpressionSyntax)node;
+                            return expression.Right == previousNode;
+                        }
+                }
+
+                if (!IsValidConstantPatternSubtreeKind(kind))
+                {
+                    return false;
+                }
+
+                previousNode = node;
+                node = node.Parent;
+            }
```

Comment:
still don't know why any of this is necessary :)

feels like you just need to handle the trivial case of `x is Y.`

what cases does this not handle properly, and why can't we just special case those instead of adding this complex walk?